src/dawn/Optimizer/PassSetNonTempCaches.cpp

```diff
@@ -0,0 +1,325 @@
+        std::make_shared<StatementAccessesPair>(assignmentStatement);
+    std::shared_ptr<Accesses> newAccess = std::make_shared<Accesses>();
+    newAccess->addWriteExtent(assignmentID, Extents({0, 0, 0}));
+    newAccess->addReadExtent(assigneeID, Extents({0, 0, 0}));
+    pair->setAccesses(newAccess);
+    domethod->getStatementAccessesPairs().push_back(pair);
+
+    // Add the new expressions to the map
+    instantiation_->getExprToAccessIDMap().emplace(fa_assignment, assignmentID);
+    instantiation_->getExprToAccessIDMap().emplace(fa_assignee, assigneeID);
+  }
+
+  /// @brief checks if there is a read operation before the first write operation in the given
+  /// multistage.
+  /// @param[in]    id  original FieldAccessID of the field to check
+  /// @return           true if there is a read-access before the first write access
+  bool checkReadBeforeWrite(int id) {
+    for(auto stageItGlob = multiStagePrt_->getStages().begin();
+        stageItGlob != multiStagePrt_->getStages().end(); ++stageItGlob) {
+      DoMethod& doMethod = (**stageItGlob).getSingleDoMethod();
+      for(int stmtIndex = 0; stmtIndex < doMethod.getStatementAccessesPairs().size(); ++stmtIndex) {
+        const std::shared_ptr<StatementAccessesPair>& stmtAccessesPair =
+            doMethod.getStatementAccessesPairs()[stmtIndex];
+
+        // Find first if this statement has a read
+        auto readAccessIterator = stmtAccessesPair->getCallerAccesses()->getReadAccesses().find(id);
+        if(readAccessIterator != stmtAccessesPair->getCallerAccesses()->getReadAccesses().end()) {
+          return true;
+        }
+        // If we did not find a read statement so far, we have  a write first and do not need to
+        // fill the cache
+        auto wirteAccessIterator =
+            stmtAccessesPair->getCallerAccesses()->getWriteAccesses().find(id);
+        if(wirteAccessIterator != stmtAccessesPair->getCallerAccesses()->getWriteAccesses().end()) {
+          return false;
+        }
+      }
+    }
+    return false;
+  }
+
+  bool checkReadOnlyAccess(int id) {
+      for(auto stageItGlob = multiStagePrt_->getStages().begin();
+          stageItGlob != multiStagePrt_->getStages().end(); ++stageItGlob) {
+        DoMethod& doMethod = (**stageItGlob).getSingleDoMethod();
+        for(int stmtIndex = 0; stmtIndex < doMethod.getStatementAccessesPairs().size(); ++stmtIndex) {
+          const std::shared_ptr<StatementAccessesPair>& stmtAccessesPair =
+              doMethod.getStatementAccessesPairs()[stmtIndex];
+
+          // If we find a write-statement we exit
+          auto wirteAccessIterator =
+              stmtAccessesPair->getCallerAccesses()->getWriteAccesses().find(id);
+          if(wirteAccessIterator != stmtAccessesPair->getCallerAccesses()->getWriteAccesses().end()) {
+            return false;
+          }
+        }
+      }
+      return true;
+  }
+
+  MultiStage* multiStagePrt_;
+  HardwareConfig config_;
```

Comment:
This is never initialized and just relies on default initialization. We should open an issue that the `HardwareConfig` gets properly passed around in the Passes. One way would be to extend the `Pass::run()` to take an additional argument i.e the `HardwareConfig`

Reply:
Should it be moved around though? Should this not in the end be a singleton that is in a place where it is globally accessible for all passes?

Reply:
Singletons in library is usually a bad design, I would argue you have to pass it around. You can also simply attach it to the `StencilInstantiation` i.e ISIR

Reply:
ISIR seems like an ok place to put it. This allows for a very simple pass to set it properly. I see two problems with this though: 
- The first one is that we add an additional field to the ISIR which is not really an ISIR feature?
- This is not something that is not really dependent on a Stencil/Stencil-Function but is the same thing on a global scope. 

I'd suggest to add it to the optimizer.

Thoughts?